rladmin reference (Doc-83, DOC-210, DOC-606) (#1256)
RedisLabs/redislabs-docs: content/rs/references/rladmin.md

> ---

Reviewer: General comment - @AlonMagrafta @kaitlynmichaelredislabs thoughts about how we indicate which commands are `dangerous` - could lead to data loss or `caution` - can have other adverse impacts - if at all?
Some of these are required for setup and regular maintenance while some are advanced and shouldn't be used without consulting Redis Labs support.

Reviewer 2: I think we should mention any potential danger of data loss. I just need to know which commands are "dangerous" and how data loss or any other potential pitfalls can be avoided.

Reviewer 3: It's a good idea. We will need a thorough review of it to be able to mark them all

Author: I created DOC-618 for marking the "dangerous" commands

| [recover](#recover) | Recover databases |
| [restart](#restart) | Restart database shards |
| [status](#status) | Show status information |
| [suffix](#suffix) | Suffix management |
| [tune](#tune) | Tune system parameters |
| [upgrade](#upgrade) | Upgrade the Redis or module version for a database |
| [verify](#verify) | Cluster verification reports |

### `bind`

`rladmin bind` manages the proxy policy for the cluster or a specific database. It also can include or exclude endpoints from the proxy policy.

```text
rladmin bind 
        [ db <db:id | name> ] 
        endpoint <id> 
        policy <single | all-master-shards | all-nodes>
```

| Proxy policy | Description |
| - | - |
| single | All traffic flows through a single proxy bound to the database endpoint (preferable in most cases)               |
| all-master-shards | Multiple proxies, one on each master node (best for high traffic and multiple master shards)                     |
| all-nodes | Multiple proxies, one on each node of the cluster (increases traffic in the cluster, only used in special cases) |
| Legacy | Copies existing binding configuration from earlier versions

> 

Author: I am going to push out the initial version of this page today and add links the next revision.

rladmin cluster recover 
        filename <recovery-file-name> 
        [ ephemeral_path <path> ] 
        [ persistent_path <path> ] 
        [ rack_id <ID> ] 
        [ override_rack_id ] 
        [ node_uid <number> ] 
        [ flash_enabled ] 
        [ flash_path <path> ] 
        [ addr <ip-address> ] 
        [ external_addr <ip-addresses> ]
```

| Optional Parameters | Description |
| - | - |
| ephemeral_path | Path to ephemeral storage location |
| persistent_path | Path to persistent storage location |
| rack_id | Rack ID of the rack |
| override_rack_id | Manually overrides the existing default rack ID |
| flash_enabled | Enables flash storage |
| flash_path | Path to flash storage location |
| addr | Internal IP addresses of the node |
| external_addr | External IP addresses of the node |

### `failover`

`rladmin failover` performs a failover for master shards of a specific database, and promotes their respective replicas to new masters. The `immediate` flag performs the failover without verifying the replica shards are in full sync with the master shards.

```text
rladmin failover 
        [ db <db:id | name> ] 
        shard <id1 .. idN> 
        [ immediate ]
```

> ### `help`

Reviewer: @kaitlynmichaelredislabs If I may suggest - can we start with the Help and Info commands and then move up in order of complexity/impact/danger?

Reviewer 2: That's an interesting question. If they are organized by difficulty and not alphabetically, this page might be more difficult to navigate. Do we know if there is an industry standard for how to organize command reference material?

Reviewer: Here are a couple of references I'm familiar with and I think are pretty well organized.
https://docs.microsoft.com/en-us/cli/azure/get-started-with-azure-cli
https://docs.openshift.com/container-platform/4.7/cli_reference/openshift_cli/getting-started-cli.html

I really like how the Azure guide explains what it is, getting started, and then points to popular uses. Maybe we need to separate these from the alphabetized reference guide.

`rladmin help` list all the options and parameters associated with the `rladmin` command.

### `info`

`rladmin info` lists configurable settings for specified databases, proxies, clusters, or nodes.

```text
rladmin info db <db:id | name>
```

```text
rladmin info proxy <id | all>
```

```text

> rladmin info cluster <id>

Reviewer: ```suggestion
rladmin info cluster
```
I don't think `rladmin info cluster` accepts additional parameters

-  `cloud` profile is suitable for common cloud environments. It has a higher tolerance for network jitter.
- `local-network`profile is suitable for dedicated LANs and has a better failure detection and failover times.

### `upgrade`

`rladmin upgrade` schedules a restart of the master and slave processes to upgrade the Redis software for a specific database instance.

```text
rladmin upgrade db <db:id | name> 
        [ preserve_roles ] 
        [ keep_current_version ] 
        [ discard_data ] 
        [ force_discard ]
```

| Optional Parameters | Description |
| - | - |
| preserve_roles | Performs an additional failover to guarantee roles of masters are preserved |
| keep_current_version | Upgrades to a new patch release, not to the latest major.minor version |
| discard_data | Indicates that data will not be saved after the upgrade |
| force_discard | Forces `discard_data` if there is replication or persistence enabled |

### `verify`

`rladmin verify balance` prints a cluster shard balance report.

```text
rladmin verify balance [ node <id> ]
```

`rladmin verify rack_aware` prints a cluster rack aware verification report.

```text
rladmin verify rack_aware
```

> ## Use the rladmin shell

Reviewer: I would lead with this info.

Reviewer 2: I don't think the rladmin shell adds much value over just running individual rladmin commands besides the fact that it saves you a few keystrokes. Are there any capabilities the rladmin shell gives users that I missed in the list below?

Reviewer: I think that it is much easier and user-friendly to run from the shell, especially for beginner users that may not even know that it is an option.

Author: I will make sure to mention the shell when writing "how-to" articles. When I can expand each section into its own page, we will be able to add more detail.

---
Title: rladmin
description:
weight: $weight
alwaysopen: false
categories: ["RS"]
aliases: /rs/references/cli-reference/rladmin/
---

> `rladmin` is a command-line utility for performing administrative tasks such as failover, migrate, and bind on a Redis Enterprise Software cluster. Some of these tasks can also be performed through the Management UI and some are unique to the `rladmin` CLI tool.

Reviewer 3: ```suggestion
`rladmin` is a command-line utility for performing administrative tasks such as failover, migrate, and bind an endpoint on a Redis Enterprise Software cluster as well as cluster and database/s configuration updates. Some of these tasks can also be performed through the Management UI and some are unique to the `rladmin` CLI tool.
```

Author: ```suggestion
`rladmin` is a command-line utility for performing administrative tasks such as failover, migration, and endpoint binding on a Redis Enterprise Software cluster. `rladmin` can also edit cluster and database configurations. Some of these tasks can also be performed through the Management UI and some are unique to the `rladmin` CLI tool.
```

## rladmin Commands

| Command | Description |
| - | - |

> | [rladmin](#use-the-rladmin-shell) | Enter the `rladmin` shell |

Author: These options are mentioned in the rladmin shell section. I want to keep this table as clean as possible.

| [bind](#bind) | Bind an endpoint |
| [cluster](#cluster) | Cluster management commands |
| [exit](#use-the-rladmin-shell)| Exit `rladmin` shell |
| [failover](#failover) | Fail over master to replica |
| [help](#help) | Show available commands or specific command usage |
| [info](#info) | Show current parameters for a cluster, database, node, or proxy |

> | [migrate](#migrate) | Migrate elements between nodes |

Author: From the info I have, it looks like `rladmin migrate` only migrates shards and endpoints (not databases or nodes)

| all-nodes | Multiple proxies, one on each node of the cluster (increases traffic in the cluster, only used in special cases) |
| Legacy | Copies existing binding configuration from earlier versions

```text
rladmin bind 
        [ db <db:id | name> ] 
        endpoint <id> [ include | exclude ] 
        proxy <id1 .. idN>
```

### `cluster`

`rladmin cluster` manages cluster configuration and administration. Some `rladmin cluster` commands are only for clusters that are already configured, while others are only for new clusters that have not been configured.

| Command (for configured clusters) | Description |
| - | - |
| `cluster certificate` | Sets the cluster certificate |
| [`cluster config`](#cluster-config) | Updates configuration for the cluster |
| [`cluster reset_password`](#cluster-reset_password) | Changes the password for a given email |
| [`cluster stats_archiver`](#cluster-stats_archiver) | Enables/disables stats archiving |
| [`cluster debug_info`](#cluster-debug_info) | Creates a support package |
| `cluster running_actions` | Lists all active tasks |

> | Command (for non-configured clusters) | Description |

Reviewer 3: I advise to move all these commands together with their extended description (currently lines 149-257) to a separate dedicated section about uninitialized clusters/nodes

Author: I'll look into this change with DOC-617

| - | - |
| [`cluster create`](#cluster-create) | Creates a new cluster |
| [`cluster join`](#cluster-join) | Adds a node to an existing cluster |
| [`cluster recover_filename`](#cluster-recover_filename) | Recovers a cluster from a backup file |

#### `cluster certificate`

>  `rladmin cluster certificate` sets the cluster certificate to a given file.

Author: This will be for DOC-616

 ```text
 rladmin cluster certificate 
        set <certificate_name> 
        certificate_file <certificate_file> 
        [ key_file <certificate_file> ]
```

#### `cluster config`

`rladmin cluster config` updates the cluster configuration.

```text

>  rladmin cluster config 

Author: @AlonMagrafta  Do you have the arguments and descriptions for `sentinel_cipher_suites`, `services`, `upgrade_mode`?

```text
rladmin cluster recover 
        filename <recovery-file-name> 
        [ ephemeral_path <path> ] 
        [ persistent_path <path> ]
        [ ccs_persistent_path <path> ]
        [ ccs_persistent_path <path> ]
        [ rack_id <ID> ] 
        [ override_rack_id ] 
        [ node_uid <number> ] 
        [ flash_enabled ] 
        [ flash_path <path> ] 
        [ addr <ip-address> ] 
        [ external_addr <ip-addresses> ]
```

| Optional Parameters | Description |
| - | - |
| ephemeral_path | Path to ephemeral storage location (defaults to /var/opt/redislabs) |
| persistent_path | Path to persistent storage location (defaults to /var/opt/redislabs/persist) |
| ccs_persistent_path | Path to location of ccs snapshots (default is same as persistent_path) |
| rack_id | Rack ID of the rack |
| override_rack_id | Manually overrides the existing default rack ID |
| flash_enabled | Enables flash storage |
| flash_path | Path to flash storage location (defaults to /var/opt/redislabs/flash) |
| addr | Internal IP addresses of the node |
| external_addr | External IP addresses of the node |

#### `cluster services`

> `rladmin cluster services` enables or disables selected cluster services.

Reviewer 3: https://docs.redislabs.com/latest/rs/administering/troubleshooting/disabling-services/

Author: I will add links during the DOC-616 version

```text
rladmin cluster services [ cm_server | crdb_coordinator | crdb_worker | mdns_server | pdns_server | saslauthd | stats_archiver <enabled | disabled> ]
```

> ### `failover`

Reviewer 3: @AlonMagrafta  : Mark to myself, I reviewed this PR to this point, not including the extended part of cluster create+recover+join
